go-controller/pkg/node/gateway_shared_intf.go

```diff
@@ -916,6 +927,144 @@ func (npw *nodePortWatcher) UpdateService(old, new *corev1.Service) error {
+
+	deletedCount = extIPsDeletedCount + clusterIPsDeletedCount + nodePortDeletedCount
+	errs = append(errs, extIPsErrs...)
+	errs = append(errs, clusterIPsErrs...)
+	errs = append(errs, nodePortErrs...)
+
+	if deletedCount > 0 {
+		klog.Infof("Deleted %d UDP conntrack entries for service %s/%s", deletedCount, old.Namespace, old.Name)
+	}
+
+	return utilerrors.Join(errs...)
+}
+
+// IPStringer is a constraint for types that can be converted to string IP representations
+type IPStringer interface {
+	~string | net.IP
+}
+
+// ipToString converts various IP types to string
+func ipToString[T IPStringer](ip T) string {
+	switch v := any(ip).(type) {
+	case string:
+		return v
+	case net.IP:
+		return v.String()
+	default:
+		return ""
+	}
+}
+
+// JoinIPServicePort returns the IP:port:targetPort:nodePort combination of a service port for
+// a NodePort type service, otherwise returns IP:port:targetPort combination of the service port.
+func JoinIPServicePort(ip string, servicePort corev1.ServicePort, isNodePort bool) string {
+	joinString := fmt.Sprintf("%s:%d:%s", ip, servicePort.Port, servicePort.TargetPort.String())
+	if isNodePort {
+		joinString = fmt.Sprintf("%s:%d", joinString, servicePort.NodePort)
+	}
+	return joinString
+}
+
+// deleteStaleUDPConntrackForUDPPortsAndIPs deletes the conntrack entries for the updated UDP ports of the service.
+// The conntrack entries are deleted for the old IP:port:targetPort combinations (IP:port:targetPort:nodePort for
+// NodePort type service) that are not present in the corresponding new combinations.
+func deleteStaleUDPConntrackForUDPPortsAndIPs[T IPStringer](
+	newServicePorts []corev1.ServicePort,
+	newIPs []T,
+	oldServicePorts []corev1.ServicePort,
+	oldIPs []T,
+	namespace,
+	name string,
+	isNodePort bool,
+) (int, []error) {
+	var errs []error
+	deletedCount := 0
+
+	// Get the new IP:port:targetPort combinations (IP:port:targetPort:nodePort for NodePort type service)
+	newIPPortTargetPorts := sets.New[string]()
+	for _, newPort := range newServicePorts {
+		if newPort.Protocol != corev1.ProtocolUDP {
+			continue
+		}
+		for _, newIP := range newIPs {
+			newIPPortTargetPorts.Insert(JoinIPServicePort(ipToString(newIP), newPort, isNodePort))
+		}
+	}
+
+	// Delete the conntrack entries for the old IP:port:targetPort combinations (IP:port:targetPort:nodePort for
+	// NodePort type service) that are not present in the corresponding new combinations
+	for _, oldPort := range oldServicePorts {
+		if oldPort.Protocol != corev1.ProtocolUDP {
+			continue
+		}
+		port := oldPort.Port
+		ipType := "svcVIP"
+		portType := "svcPort"
+		if isNodePort {
+			port = oldPort.NodePort
+			ipType = "nodeIP"
+			portType = "nodePort"
+		}
+		for _, oldIP := range oldIPs {
+			if !newIPPortTargetPorts.Has(JoinIPServicePort(ipToString(oldIP), oldPort, isNodePort)) {
+				if err := util.DeleteConntrackServicePort(ipToString(oldIP), port, oldPort.Protocol,
```

Comment:
We have some inconsistent logging when calling `DeleteConntrackServicePort`. [On some places we log](https://github.com/arkadeepsen/ovn-kubernetes/blob/b43a212c4f467c91595917146be582153e26b998/go-controller/pkg/node/default_node_network_controller.go#L1375), here we don't.